CRM/MembersOnlyEvent/Hook/PageRun/Register.php

```diff
@@ -0,0 +1,172 @@
+<?php
+
+use CRM_MembersOnlyEvent_Hook_PageRun_Base as Base;
+use CRM_MembersOnlyEvent_BAO_MembersOnlyEvent as MembersOnlyEvent;
+
+/**
+ * Class for Event Info PageRun Hook
+ */
+class CRM_MembersOnlyEvent_Hook_PageRun_Register extends Base {
+
+  use CRM_MembersOnlyEvent_Trait_userHasEventAccess;
+
+  /**
+   * Checks if the hook should be handled.
+   *
+   * @param $pageName
+   * @param $page
+   *
+   * @return bool
+   */
+  protected function shouldHandle($pageName, &$page) {
+    if ($pageName === CRM_Event_Page_EventInfo::class) {
+      return TRUE;
+    }
+    return FALSE;
+  }
+
+  /**
+   * Callback for event info page
+   *
+   * @param $page
+   */
+  protected function pageRun(&$page) {
+    $eventID = $page->_id;
+
+    $this->showSessionMessageWhenRegisteringAnotherParticipant($eventID);
+
+    $membersOnlyEvent = MembersOnlyEvent::getMembersOnlyEvent($eventID);
+    $userHasEventAccess = $this->userHasEventAccess($membersOnlyEvent);
+    if ($userHasEventAccess) {
+      // skip early and show the page if the user has access to the members-only event.
+      return;
+    }
+
+    $this->hideEventInfoPageRegisterButton();
+
+    $this->handleAccessOptionForUser($eventID);
+  }
+
+  /**
+   * Handle session message if the user is trying
+   * to register another participant.
+   *
+   * @param int $eventID
+   *
+   */
+  private function showSessionMessageWhenRegisteringAnotherParticipant($eventID) {
+    $isEventForMembersOnly = MembersOnlyEvent::getMembersOnlyEvent($eventID);
+
+    if (!$isEventForMembersOnly) {
+      return;
+    }
+
+    $session = CRM_Core_Session::singleton();
+    $statusMessages = $session->get('status');
+    if (empty($statusMessages)) {
+      return;
+    }
+
+    foreach ($statusMessages as $k => $msg) {
+      if (strpos($msg['text'], 'register another participant')) {
+        $statusMessages[$k]['text'] = ts("It looks like you are already registered for this event. If you want to change your registration, or you feel that you've gotten this message in error, please contact the site administrator.");
+      }
+    }
+    $session->set('status', $statusMessages);
+  }
+
+  /**
+   * Hides the event info page action links which contain
+   * the event register link.
+   */
+  private function hideEventInfoPageRegisterButton() {
+    CRM_Core_Region::instance('event-page-eventinfo-actionlinks-top')
+      ->update('default', [
+        'disabled' => TRUE,
+      ]);
+    CRM_Core_Region::instance('event-page-eventinfo-actionlinks-bottom')
+      ->update('default', [
+        'disabled' => TRUE,
+      ]);
+  }
+
+  /**
+   * Handles access options for logged / anonymous user.
+   *
+   * @param $eventID
+   *
+   */
+  private function handleAccessOptionForUser($eventID) {
+    $membersOnlyEvent = MembersOnlyEvent::getMembersOnlyEvent($eventID);
```

Comment:
you are calling this twice in this class

Reply:
I couldn't refactor it one go -  to make sure the old code working properly when moving from `membersonlyevent.php` - and in the next commit I fixed that.
for this one I want to  and